libraries/BufferedBleSpp/Xamarin Forms/BufferedBleSpp.cs

```diff
@@ -0,0 +1,153 @@
+                                await rxBufferCharacteristic.WriteAsync(outBuffer, ctsToken);
+                                remaining -= bytesRead;
+
+
+
+                                #region *// Update progress
+                                toTransmitCount += bytesRead;
+                                var progress = 0.5f * toTransmitCount / toTransmitAmount;
+                                onProgress?.Invoke(progress); 
+                                #endregion
+                            }
+                        }
+                    }
+                    #endregion
+                    #endregion
+
+                    #region *// Receive data
+                    #region *// Wait for server to respond with the data amount to receive
+                    // await DoEvents();
+                    var result = await txAmountCharacteristic.ReadAsync(ctsToken);
+                    var dataAmount = BitConverter.ToInt32(result, 0);
+                    var dataCount = 0;
+                    while ((dataAmount == -1) && (isCancelled == false))
+                    {
+                        result = await txAmountCharacteristic.ReadAsync(ctsToken);
+                        dataAmount = BitConverter.ToInt32(result, 0);
+                    }
+                    if (isCancelled) throw new Exception("The SendMessage command has been cancelled. Probably due to timeout");
+                    #endregion
+
+                    #region *// Read data chunks from server until all consumed
+                    txByteArray = new byte[0];
+                    while ((dataCount != dataAmount) && (isCancelled == false))
+                    {
+                        var data = await txBufferCharacteristic.ReadAsync(ctsToken); // Slow, but reliable
+                        txByteArray = Combine(txByteArray, data);
+                        dataCount += data.Length;
+
+                        #region *// Update progress
+                        var progress = (0.5f * dataCount / dataAmount) + 0.5f;
+                        onProgress?.Invoke(progress); 
+                        #endregion
+                    }
+                    if (isCancelled) throw new Exception("The SendMessage command has been cancelled. Probably due to timeout");
+                    #endregion
+                    tcs.SetResult(txByteArray);
+                    #endregion
+                }
+                catch (Exception ex)
+                {
+                    tcs.SetException(ex);
+                }
+            });
+            return tcs.Task;
+
+            // Helper
+            byte[] Combine(byte[] first, byte[] second)
+            {
+                byte[] bytes = new byte[first.Length + second.Length];
+                Array.Copy(first, 0, bytes, 0, first.Length);
+                Array.Copy(second, 0, bytes, first.Length, second.Length);
+                return bytes;
+            }
+        }
+    }
+}

```

Comment:
I cannot comment the next part, but please do not provide the zip.
But best is to provide a sln with csproj, something that can build ;-)